### DISCUSSION TITLE
Nested PMC references

### DISCUSSION THREAD

**Discussion by ThatCanadianGuy** [kibook/s1kd-tools#11] [Q&A]
Hi,

I'm currently dealing with a very large S1000D project that contains a folder structure like the following:

- PMC File that references other PMC files
    - Folder of DMC xml files
    - Folder of DMC xml files
    - Folder with a PMC and many DMC files
        - PMC file at root
        - Folder with DMC files

My goal would be to flatten the whole thing to XML using a variation of the  Build.sh file packaged in the 2PDF project but then convert it to HTML using XSL. My current Build.sh looks like:

```
#!/bin/sh

# Usage: sh build.sh <OUTDIR> <CSDB> <PM> <MEDIA> <VERSION>

set -e

if test "$#" -ne 5
then
	echo "Usage: sh build.sh <OUTDIR> <CSDB> <PM> <MEDIA>"
	echo
	echo "Options:"
	echo "  <OUTDIR>   Directory to write filtered objects to."
	echo "  <CSDB>     Directory to read objects from."
	echo "  <PM>       The publication module to use."
	echo "  <MEDIA>    The type of media (web or pdf)."
	echo "  <VERSION>  Version of the product."
	echo
	echo "Example:"
	echo "  sh build.sh custom csdb PMC-XX-12345-00001-00 pdf"
	exit 1
fi

OUTDIR="$1"
CSDB="$2"
PM="$3"
MEDIA="$4"
VERSION="$5"

# Create the output directory
rm -rf "$OUTDIR"
mkdir "$OUTDIR"
echo $PM
# Filter the PM and DMs and copy them to the output directory
s1kd-refs -r -DPs -d "$CSDB" $PM --verbose | s1kd-instance -f -r -L -O "$OUTDIR" -s media:prodattr="$MEDIA" -Y "Version: $VERSION" --verbose

# s1kd-refs -Ds -d csdb $(csdb/PMC-*.xml) | s1kd-instance -L -O "custom" -s media:prodattr="web" -Y "Version: 3.3.0"

# Copy the ICNs used by the filtered PM and DMs to the output directory
s1kd-ls -r -DP "$OUTDIR" | s1kd-refs -r -Gl -d "$CSDB" --verbose | sort -u | xargs -d '\n' cp -t "$OUTDIR"

# Synchronize references in the filtered DMs
s1kd-ls -r -D "$OUTDIR" | s1kd-syncrefs -fl --verbose

# Validate the filtered PM and DMs
s1kd-ls -DP "$OUTDIR" | s1kd-validate -v -l --net
s1kd-ls -DP "$OUTDIR" | s1kd-brexcheck -L -p
```

Finishing it off in my make file with:
```
END_SOURCE.xml: custom
	s1kd-ls -DP $< | xargs s1kd-flatten -v -p | xml-format -o $@
```

All of this seems to generally go through. The custom folder contains all the DMC, PMC, ICN files I would normally expect, but the xml-format from the above returns:
```
$ s1kd-ls -DP "custom" | xargs s1kd-flatten -v -p | xml-format -o END_SOURCE.xml
-:125702: parser error : XML declaration allowed only at the start of the document
<?xml version="1.0"?>
     ^
-:125703: parser error : Extra content at the end of the document
<publication><dmodule xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmln
^
xargs: s1kd-flatten: terminated by signal 13
```

The error seems to indicate bad source XML, but it seems off to me. Am I missing any steps for this process? Any help would be greatly appreciated.

**Comment by kibook** [comment]
I think this is an issue with xargs. xargs has a limit to the number of arguments it will take, and if it receives more, it will split the argument list between multiple calls to the specified program. This means you end up getting multiple separate output documents from s1kd-flatten.

You can tweak the behaviour of xargs with options like --max-args for the maximum number of arguments, if you know roughly the number of DMs and PMs you have. But for a more flexible solution, you'd want to use the --list option that many of the s1kd-tools have, which makes it read arguments from stdin similar to xargs, but without any limit (aside from available computer memory of course). However, it seems the s1kd-flatten tool was never given this option. I'm going to look into this and hopefully implement a --list option for s1kd-flatten soon.

**Comment by kibook** [comment]
Added in d948cad. You should be able to change your Makefile to this:
```Makefile
END_SOURCE.xml: custom
	s1kd-ls -DP $< | s1kd-flatten -l -v -p | xml-format -o $@
```

**Reply by ThatCanadianGuy** [reply]
Thanks! I actually ended up opting for a PowerShell script path so I translated the bash/makefile into a single PowerShell script. That said, I'm running into some challenges in the front matter world. I've been able to generate a ToC, LoEDM or Highlights, but my next step would be using xsl to turn it into HTML. I was thinking for ToC I would use `<ol>` with recursive list items/ other `<ol>s` as it recursed and for the LoEDM I would probably create a table similar to your PDF example (similar for Highlights). 

Before I started to dig into that, I wanted to ask if you had XSL files that converted front matter into HTML already that I may have missed? I'm not an expert in the subject so any info would be greatly appreciated! Thanks for all your help already as well.
